### ISSUE TITLE
Nested Text with onPress

### ISSUE THREAD

**Issue by zackify** [facebook/react-native#27549]
Hey there. I found an issue when rendering nested Text elements. It's almost the exact same as this ticket: https://github.com/facebook/react-native/issues/1030

I was able to get it to sort of work. I had to add an `onPress` to the Text component.

Problems:
- TypeScript says there isn't an `onPress` on Text elements. But it does in fact work. This should probably be fixed in the type definitons.
- Not possible to use TouchableOpacity, so it doesn't feel good when pressing on these items. 

When using TouchableOpacity like this:

```js
<Text>
  <Text>first part</Text>
  <TouchableOpacity><Text>Second part</Text></TouchableOpacity>
</Text>
```
Second part doesn't get rendered at all.

Before you suggest using a `<View>` around the `<Text>` instead, please look at the referenced issue. When you do that, the text runs off screen, or wraps weirdly.

TLDR; I need to add a touchable opacity inside a nested Text component. Our api returns text in blocks, the RN app needs to parse it and render an array of text elements together with different styling.

React Native version:

0.61.4

**Comment by zackify** [comment]
Still a problem

**Comment by darrylyoung** [comment]
Yeah, it'd be great if this was possible as I just came across the same issue when trying to nest links inside a string of translated text.

Right now, I'm having to use `onPress` on my `Text` component and by using that instead of `TouchableOpacity`, I get no control over `hitSlop`, `activeOpacity`, etc. It just gives me an ugly grey background to the text when I press it, which I can only turn off with `suppressHighlighting`.

**Comment by backmeupplz** [comment]
FYI: when nesting just `Text` tags on iOS the gestures get propagated :) So this works:

```javascript
<Text onLongPress={() => console.log('1')}>
  <Text>
    <Text>Some text </Text>
    <Text onPress={() => console.log('2')}>clickable text</Text>
    <Text> another text</Text>
  </Text>
</Text>
```

**Comment by backmeupplz** [comment]
Any activity on it? 🤔

**Comment by vinaysharma14** [comment]
Hey Zack, this should help you out.

```
<View style={{ flexDirection: 'row' }}>
    <Text>first part </Text>
    <TouchableOpacity>
      <Text>second part</Text>
    </TouchableOpacity>
</View>
```

**Comment by vinaysharma14** [comment]
Hey @zackify, this should help you out.

```
<View style={{ flexDirection: 'row' }}>
  <Text>first part </Text>
  <TouchableOpacity><Text>second part</Text></TouchableOpacity>
</View>
```

**Comment by vinaysharma14** [comment]
Hey @zackify, this should help you out :)

```
<View style={{ flexDirection: 'row' }}>
    <Text>first part </Text> // notice the empty space space after part
    <TouchableOpacity>
        <Text>second part</Text>
    </TouchableOpacity>
</View>
```

**Comment by vinaysharma14** [comment]
Hey @zackify, this should help you out :)

```javascript
<View style={{ flexDirection: 'row' }}>
    <Text>first part </Text> // notice the empty space space after part
    <TouchableOpacity>
        <Text>second part</Text>
    </TouchableOpacity>
</View>
```

**Comment by vinaysharma14** [comment]
Hey @zackify, this should help you out :)

```
javascript
<View style={{ flexDirection: 'row' }}>
    <Text>first part </Text> // notice the empty space space after part

    <TouchableOpacity>
        <Text>second part</Text>
    </TouchableOpacity>
</View>
```

**Comment by vinaysharma14** [comment]
Hey @zackify, this should help you out :)

```javascript
<View style={{ flexDirection: 'row' }}>
    <Text>first part </Text> // notice the empty space space after part

    <TouchableOpacity>
        <Text>second part</Text>
    </TouchableOpacity>
</View>
```

**Comment by backmeupplz** [comment]
@vinaysharma14 looks like the second part still doesn't render. Are you sure it works? What RN version are you running?

**Comment by backmeupplz** [comment]
Not fixed yet!

**Comment by Stevemoretz** [comment]
Great we can't do it yet :)

**Comment by vinaysharma14** [comment]
> @vinaysharma14 looks like the wrapping is broken when you do it like this :(

Hi @zackify @backmeupplz @Stevemoretz I guess this should work fine :)

```
<Text>
  <Text>
    I believe Nikita had said that sentence wrapping would break{' '}
  </Text>

    <TouchableOpacity>
    <Text>here</Text>
  </TouchableOpacity>

  <Text>
    {' '}but it didn’t it!
  </Text>
</Text>
```

**Comment by vinaysharma14** [comment]
> @vinaysharma14 looks like the wrapping is broken when you do it like this :(

Hi @zackify @backmeupplz @Stevemoretz I guess this should work fine :)

```
<Text>
  <Text>
    I believe @backmeupplz had said that sentence wrapping would break{' '}
  </Text>

   <TouchableOpacity>
    <Text>here</Text>
  </TouchableOpacity>

  <Text>
    {' '}but it didn’t it!
  </Text>
</Text>
```

**Comment by vinaysharma14** [comment]
> @vinaysharma14 looks like the wrapping is broken when you do it like this :(

Hi @zackify @backmeupplz @Stevemoretz I guess this should work fine :)

```
<Text>
  <Text>
    I believe @backmeupplz had said that sentence wrapping would break{' '}
  </Text>

  <TouchableOpacity>
    <Text>here</Text>
  </TouchableOpacity>

  <Text>
    {' '}but it didn’t it!
  </Text>
</Text>
```

**Comment by backmeupplz** [comment]
@vinaysharma14 does this work on iOS though?

```typescript
<TouchableOpacity onLongPress={() => console.log('1')}>
  <Text>
    <Text>Some text </Text>
    <Text onPress={() => console.log('2')}>clickable text</Text>
    <Text> another text</Text>
  </Text>
</TouchableOpacity>```

**Comment by vinaysharma14** [comment]
@backmeupplz I didn’t try but it's not OP's requirement.

**Comment by vinaysharma14** [comment]
@backmeupplz I didn’t try but it's not OP's requirement.

**Comment by backmeupplz** [comment]
@vinaysharma14 sure, I can create a separate issue for this

**Comment by vinaysharma14** [comment]
@backmeupplz may I know what functionality are you trying to achieve with this code snippet? I'm unable to understand the usecase of adding touchable on entire text and an onPress on a word.

**Comment by backmeupplz** [comment]
@vinaysharma14 sure. It's a text of a todo in a list of todos. Todo text has links in it. Tapping the text in general copies the whole text. Taping the links opens links.

**Comment by vinaysharma14** [comment]
@backmeupplz I visualised the code you wrote into a wireframe. Can you confirm if this is the desired behaviour?

<img width="780" alt="Screenshot 2021-06-13 at 1 47 58 AM" src="https://user-images.githubusercontent.com/48080660/121788236-65c28f00-cbe9-11eb-9a13-fd8d0ef7c798.png">

**Comment by backmeupplz** [comment]
@vinaysharma14 this looks correct, yes. I can't remember if I had a long tap on the outer side or just a single tap, it was a while ago — but general idea is correct

**Comment by vinaysharma14** [comment]
@backmeupplz thanks for confirming that. Either it's a long or single tap, `TouchableOpacity` is sufficient. So, if we are able to get it working, both taps would work. Are you seeking a solution to this currently? You can give a thumbs up and I can check how we can have it implemented in React Native.

**Comment by vinaysharma14** [comment]
@backmeupplz thanks for confirming that. Either it's a long or single tap, `TouchableOpacity` is sufficient. So, if we are able to get it working, both taps would work. If you are seeking a solution to this, you can give a thumbs up and I can check if we can implement it in React Native.

**Comment by vinaysharma14** [comment]
Hi @backmeupplz, you may check the implementation below.

### GIF Demo
![GIF Demo](https://user-images.githubusercontent.com/48080660/122112248-3a84ad80-ce3e-11eb-8869-ecb7516a2654.gif)

### Code
```js
import React, { useCallback, Children } from 'react';

import {
  View,
  Linking,
  StyleSheet,
  SafeAreaView,
  Text as RNText,
  TouchableOpacity,
} from 'react-native';

import RNClipboard from '@react-native-clipboard/clipboard';

// =================== Reusable Components =================== //

const Text = ({ children, style = {} }) => (
  <RNText style={[styles.text, style]}>{children}</RNText>
);

const Link = ({ text, link }) => {
  const openLink = useCallback(() => {
    Linking.openURL(link);
  }, [link]);

  return (
    <TouchableOpacity style={styles.touchable} onPress={openLink}>
      <Text style={styles.red}>{text}</Text>
    </TouchableOpacity>
  );
};

const Clipboard = ({ text, children }) => {
  const copyToClipboard = useCallback(() => {
    RNClipboard.setString(text);
  }, [text]);

  return (
    <TouchableOpacity onPress={copyToClipboard}>{children}</TouchableOpacity>
  );
};

// ===================== Util Function ===================== //

const isString = value => typeof value === 'string';

// ===================== Mock Links ===================== //

const { name, rapidReact, mmt, react, reactNative, node } = {
  name: {
    text: 'Vinay Sharma',
    link: 'https://www.linkedin.com/in/vinaysharma-/',
  },
  rapidReact: {
    text: 'Rapid React',
    link: 'https://www.npmjs.com/package/rapid-react',
  },
  mmt: {
    text: 'MakeMyTrip',
    link: 'https://www.makemytrip.com/',
  },
  react: {
    text: 'React',
    link: 'https://reactjs.org/',
  },
  reactNative: {
    text: 'React Native',
    link: 'https://reactnative.dev/',
  },
  node: {
    text: 'Node',
    link: 'https://nodejs.org/en/',
  },
};

// ================== Mock Message ================== //

const mockMsg = [
  'Hi, my name is ',
  name,
  '. I am the author of ',
  rapidReact,
  ' and a SDE at ',
  mmt,
  '.\n\n',
  'I love developing Full Stack applications with ',
  react,
  ', ',
  reactNative,
  ', ',
  node,
  ' and much more!',
];

const stringifiedMockMsg = mockMsg
  .map(msg => (isString(msg) ? msg : msg.text))
  .join('');

// ===================== App ===================== //

const App = () => {
  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.subContainer}>
        <Clipboard text={stringifiedMockMsg}>
          <Text>
            {Children.toArray(
              mockMsg.map(msg =>
                isString(msg) ? <Text>{msg}</Text> : <Link {...msg} />,
              ),
            )}
          </Text>
        </Clipboard>
      </View>
    </SafeAreaView>
  );
};

// ===================== Styles ===================== //

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: '#fff',
  },
  subContainer: {
    padding: 20,
    borderWidth: 1,
    borderRadius: 5,
    marginHorizontal: 50,
    borderColor: '#000',
  },
  touchable: {
    marginBottom: -3,
  },
  text: {
    fontSize: 15,
  },
  red: {
    color: 'red',
  },
});

export default App;
```

**Comment by vinaysharma14** [comment]
Hi @backmeupplz, you may check the implementation below.

### GIF Demo
![Untitled](https://user-images.githubusercontent.com/48080660/122113400-b29fa300-ce3f-11eb-9778-8a42b021446b.gif)

### Code
```js
import React, { useCallback, Children } from 'react';

import {
  View,
  Linking,
  TextInput,
  StyleSheet,
  SafeAreaView,
  Text as RNText,
  TouchableOpacity,
} from 'react-native';

import RNClipboard from '@react-native-clipboard/clipboard';

// =================== Reusable Components =================== //

const Text = ({ children, style = {} }) => (
  <RNText style={[styles.text, style]}>{children}</RNText>
);

const Link = ({ text, link }) => {
  const openLink = useCallback(() => {
    Linking.openURL(link);
  }, [link]);

  return (
    <TouchableOpacity style={styles.touchable} onPress={openLink}>
      <Text style={styles.red}>{text}</Text>
    </TouchableOpacity>
  );
};

const Clipboard = ({ text, children }) => {
  const copyToClipboard = useCallback(() => {
    RNClipboard.setString(text);
  }, [text]);

  return (
    <TouchableOpacity onPress={copyToClipboard}>{children}</TouchableOpacity>
  );
};

// ===================== Util Function ===================== //

const isString = value => typeof value === 'string';

// ===================== Mock Links ===================== //

const { name, rapidReact, mmt, react, reactNative, node } = {
  name: {
    text: 'Vinay Sharma',
    link: 'https://www.linkedin.com/in/vinaysharma-/',
  },
  rapidReact: {
    text: 'Rapid React',
    link: 'https://www.npmjs.com/package/rapid-react',
  },
  mmt: {
    text: 'MakeMyTrip',
    link: 'https://www.makemytrip.com/',
  },
  react: {
    text: 'React',
    link: 'https://reactjs.org/',
  },
  reactNative: {
    text: 'React Native',
    link: 'https://reactnative.dev/',
  },
  node: {
    text: 'Node',
    link: 'https://nodejs.org/en/',
  },
};

// ================== Mock Message ================== //

const mockMsg = [
  'Hi, my name is ',
  name,
  '. I am the author of ',
  rapidReact,
  ' and a SDE at ',
  mmt,
  '.\n\n',
  'I love developing Full Stack applications with ',
  react,
  ', ',
  reactNative,
  ', ',
  node,
  ' and much more!',
];

const stringifiedMockMsg = mockMsg
  .map(msg => (isString(msg) ? msg : msg.text))
  .join('');

// ===================== App ===================== //

const App = () => {
  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.subContainer}>
        <Clipboard text={stringifiedMockMsg}>
          <Text>
            {Children.toArray(
              mockMsg.map(msg =>
                isString(msg) ? <Text>{msg}</Text> : <Link {...msg} />,
              ),
            )}
          </Text>
        </Clipboard>
      </View>

      <TextInput
        multiline
        placeholder="Paste here"
        placeholderTextColor="#999"
        style={[styles.subContainer, styles.input]}
      />
    </SafeAreaView>
  );
};

// ===================== Styles ===================== //

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: '#fff',
  },
  subContainer: {
    padding: 20,
    borderWidth: 1,
    borderRadius: 5,
    borderColor: '#000',
    marginHorizontal: 50,
  },
  input: {
    height: 180,
    marginTop: 50,
    paddingTop: 20,
  },
  touchable: {
    marginBottom: -3,
  },
  text: {
    fontSize: 15,
  },
  red: {
    color: 'red',
  },
});

export default App;
```

### Package.json

```json
{
  "name": "foo",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint ."
  },
  "dependencies": {
    "@react-native-clipboard/clipboard": "^1.8.1",
    "@react-native-community/clipboard": "^1.5.1",
    "react": "17.0.1",
    "react-native": "0.64.2"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9",
    "@babel/runtime": "^7.12.5",
    "@react-native-community/eslint-config": "^2.0.0",
    "@types/react-native": "^0.64.10",
    "babel-jest": "^26.6.3",
    "eslint": "7.14.0",
    "jest": "^26.6.3",
    "metro-react-native-babel-preset": "^0.64.0",
    "react-test-renderer": "17.0.1"
  },
  "jest": {
    "preset": "react-native"
  }
}
```

**Comment by backmeupplz** [comment]
@vinaysharma14 thank you for such a thorough investigation and for the example! It looks like this works :) Cheers!

**Comment by devendra-learngram-ai** [comment]
Recently I encountered the same issue, as I had to render urls differently that a simple text in chat. I tried 3 different Approach, out of which the last approach worked pretty neatly and does what I want.

what I want: 
1. It should render url and plain text differently
2. It should have access to all touch events same as any Touchable components, _i.e onPressIn, onLongPress, onPressOut, and onPress_
3. It should be able to wrap multiline messages

`const MessageBox = (message) => {
  const URL_REGEX =
    /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)/;

  return (
    <Text>
      {message.split(" ").map((word) =>
        URL_REGEX.test(part) ? (
          <Text
            onResponderGrant={(event) =>
              console.log(
                "this is the time to highlight and show the user what is happening"
              )
            }
            onLongPress={() => console.log("use this to copy message")}
            onResponderRelease={(event) =>
              console.log("this is the time to open link in the browser")
            }
          >
            {url}
          </Text>
        ) : (
          <Text>{word}</Text>
        )
      )}
    </Text>
  );
};`

**Comment by aprilmintacpineda** [comment]
It's 2022 and we still can't make a proper clickable text that's inside a paragraph.

```tsx
<Text
  style={{
    marginTop: 45,
    width: 167,
    textAlign: 'center'
  }}
>
  Don’t have an account yet?{' '}
  <TouchableOpacity
    onPress={() => {
      console.log('fuck');
    }}
  >
    <Text style={{ color: 'red' }}>Create account</Text>
  </TouchableOpacity>{' '}
  now!
</Text>
```

The code above, you be able to render a paragraph, but the touchable text is not properly aligned. It will be pushed up by a few pixels and I can't find a workaround to fix it.

We have self driving cars but can't do a proper touchable text, ironic.

**Comment by Stevemoretz** [comment]
> 

Don't use touchable opacity use a Text also there it has an onPress prop, you can also modify the style for that.

**Comment by aprilmintacpineda** [comment]
> > 
> 
> Don't use touchable opacity use a Text also it has an onPress prop, you can also modify the style for that.

I know, but that won't have a proper response, I tried making my own with `Animated.Text` but the opacity won't work unless I wrap it inside View but that produces the same problem -- the text will be misaligned.

**Comment by Stevemoretz** [comment]
> @devendra-learngram-ai is there any example of how to use the onResponderGrand and onResponderRelease to imitate touchableopacity animation on text onPress?

You could try with reanimated2 Animated.Text that might actually work without any problems. Haven't tested it myself yet though it should be pretty easy to test.

**Comment by aprilmintacpineda** [comment]
> 

I'm using "react-native-reanimated": "^2.4.1" it didn't work, the `opacity` won't work until you wrap the Animated.Text inside the View. I deleted the whole code for it out of frustration, because I've been stuck here for about 1.5h searching on google for answers, I tried adding `marginTop: -3` and it worked for ios (the text was aligned) but not for android. I'll just stick to unresponsive touchable text I suppose. It's a shame but it is what it is.

**Comment by aprilmintacpineda** [comment]
Thanks for responding btw, I really appreciate it.

**Comment by Stevemoretz** [comment]
Sure! No problem.
Yeah I just remembered gestureHandler only works if there is an Animated.View nested in it, so that's not really an option here.

We at least need to have two events "on text tap" and "on text release" natively added, that way we can easily fix the issue with reanimated, right now it's impossible, (unless those events exist already that I'm not aware of)

**Comment by devendra-learngram-ai** [comment]
> @devendra-learngram-ai is there any example of how to use the onResponderGrand and onResponderRelease to imitate touchableopacity animation on text onPress?

@cakasuma I didn't want to go through extra trouble after making the Text click work, so I took another path and highlighted text on click to let user know about the click. So I set the state inside `onResponderGrant`, and based on that state I applied the background color in the `Text`. And on `ResponderRelease` I set the state back to its initial value, which will remove the background color from `Text`.

**Comment by Hostname47** [comment]
With a small addition to @aprilmintacpineda answer, you'll be able to have an aligned TouchableOpacity aligned with a text

> > 
> 
> I'm using "react-native-reanimated": "^2.4.1" it didn't work, the `opacity` won't work until you wrap the Animated.Text inside the View. I deleted the whole code for it out of frustration, because I've been stuck here for about 1.5h searching on google for answers, I tried adding `marginTop: -3` and it worked for ios (the text was aligned) but not for android. I'll just stick to unresponsive touchable text I suppose. It's a shame but it is what it is.

After hours of war, I found a workaround to fix the pixels that RN add to TouchableOpacity element when you wrap it within a Text. The key is to add a View element within your TouchableOpacity and wrap all the element within that View and add marginBottom to negative value to align it with text as shown in the following example:

```JSX
<Text style={styles.modalText}>
  You can suggest some questions along with answers by send it to us in
  <TouchableOpacity onPress={() => console.log("go to contact")}>
    <View
      style={{
        flexDirection: "row",
        alignItems: "center",
        marginHorizontal: 6,
        marginBottom: -4, // <-- This one will take care of the pixels that RN add to TouchableOpacity when you wrap it within Text
        }}
      >
        <ContactIcon width={12} height={12} fill="#08A3FD" />
        <Text
          numberOfLines={1}
          style={{
          fontWeight: "bold",
          marginLeft: 4,
          fontSize: 14,
          color: "#08A3FD",
          }}
        >
          contact
        </Text>
      </View>
    </TouchableOpacity>
    section
</Text>
```

Output : 
![312442230_386446743607075_1017777519541700690_n](https://user-images.githubusercontent.com/30189145/197538553-4afb1d08-9ea1-4645-8d06-d1a15204e0c8.jpg)

good luck <3

**Comment by whaddafish** [comment]
Issue still exists

**Comment by pfcodes** [comment]
So there really is no fix to this?

**Comment by ba9nist** [comment]
just import TouchableHighlight from 'react-native' package and not from 'react-native-gesture-handler'

**Comment by alainib** [comment]
still bugging in 2023

**Comment by maulik54e** [comment]
Still an issue

**Comment by Hemistone** [comment]
How about using `<Pressable/>`, instead?

I had similar situation in iOS for making 'KeyboardDismissView' which enables to dismiss keyboard when user touches outer area of Input, but I've solved it by implementing `<Pressable/>`, not `<TouchableWithoutFeedback/>`

```typescript
// Use <Pressable/> instead of <Touchable~~~/> Components
export default function KeyboardDismissView(props: KeyboardDismissViewProps) {
  const { children, style } = props;
  return (
    <Pressable onPress={Keyboard.dismiss} accessible={false} style={{ ...style, flexGrow: 1 }}>
      {children}
    </Pressable>
  );
};

export default function InputWithButtonScreen () {
  return(
    <KeyboardDismissView>
      <TextInput/>
      <Button onPress={()=>{console.log('1')}}>Hello</Button> // <-- Now Press Event gets in this Button, after using <Pressable/>
    </KeyboardDismissView>
  );
};
```

**Comment by Malcom-Mullen** [comment]
<PressableOpacity onPress={onPress}>
      <TouchableText
        fontFamily="Poppins-SemiBold"
        className={`text-[#D0FD3E] ${styleClassName}`}
        style={{top: Platform.OS === 'ios' ? 3 : 10}}>
        {children}
      </TouchableText>
    </PressableOpacity>

**Comment by Alexnpm** [comment]
[View my solution on Stack Overflow](https://stackoverflow.com/a/79816742/12196476)

**Comment by williamsaar** [comment]
This is still an issue in December 2025. I have researched for a quick hour and have yet to find a solution for inline links that works without sacrificing accessibility. 

- Onpress on a Text element that is inside of another Text element is not accessible to screen readers AFAIK.
- A TouchableOpacity in a Text element does not respect the text lines (see image). 

<img width="382" height="63" alt="Image" src="https://github.com/user-attachments/assets/76504115-8299-4328-8172-a0a43d961b07" />

**Comment by backmeupplz** [comment]
we are getting gta vi before this issue will be resolved it seems like
